### ISSUE TITLE
Missing aggregated usage for the running application

### ISSUE THREAD

**Issue by piotrprzybylski** [cloudfoundry-incubator/cf-abacus#88]
Application is started in August and the usage is submitted as shown below.  

```
{
  "usage": [
    {
      "start": 1439640000000,
      "end": 1439640000000,
      "region": "us-south",
      "organization_id": "c3d7fe4d-cccc-4cc3-a831-ffe98e20cf27",
      "space_id": "caeae239-f3f8-483c-9dd0-de5d41c38b6a",
      "consumer": {
        "type": "CF_APP",
        "consumer_id": "aaaaaaaf-f3f8-483c-9dd0-de5d41c38b6a"
      },
      "resource_id": "linux-container",
      "plan_id": "basic",
      "resource_instance_id": "aaaaaaaf-f3f8-483c-9dd0-de5d41c38b6a",
      "measured_usage": [
        {
          "measure": "instance_memory",
          "quantity": 536870912
        },
        {
          "measure": "running_instances",
          "quantity": 1
        }
      ]
    }
  ]
}

```

The aggregated usage for August is correct, however for September and October its empty, for example:

```
{
  "id": "k/c3d7fe4d-cccc-4cc3-a831-ffe98e20cf27/t/0001444871048322",
  "organization_id": "c3d7fe4d-cccc-4cc3-a831-ffe98e20cf27",
  "start": 1443657600000,
  "end": 1444871048322,
  "resources": [],
  "spaces": []
}
```

**Comment by jsdelfino** [comment]
Hi @piotrprzybylski 

To see usage for Sept and Oct you'll need at least one usage doc submitted during these months.

More generally, with the current design of Abacus, providers of resources metered using time-based usage metrics (like the memory consumption metric used in the linux-container sample resource) need to send at least one usage 'ping' doc per month to get the consumption from the previous month accumulated and carried over each month, while these resources are still active.

This allows inactive resources to not be carried over into each new monthly database partition and get naturally phased out (sort of garbage collected). Without that, Abacus would have to keep around usage for these inactive resources forever, causing its monthly databases to grow indefinitely.

The required 'ping' usage submission can happen anytime during the month.

While the requirement to submit that usage every month is currently on the resource provider, I'd be happy to discuss a future enhancement of the current design to have Abacus handle that carry over & phasing out automatically (based on some configuration of the particular resource for example).

**Comment by hsiliev** [comment]
The requirement to submit "ping" data every month means that the provider (in this case cf-bridge) has to access historical data about the usage. I believe not all providers keep this info and such requirement will force simple bridge-like implementations to have the old consumption data handy so they can provide the "ping" doc on month end.

Somehow all this feels against the idea of time-based usage.

In the same time the DB holding all the historical/audit events was already proposed by @piotrprzybylski, so this is not a bad thing for cf-bridge.

**Comment by jsdelfino** [comment]
@hsiliev +1 to what you said, like I said in a previous comment:
>  I'd be happy to discuss a future enhancement of the current design to have Abacus handle that carry over & phasing out automatically (based on some configuration of the particular resource for example).

We'll need to find a good way for the resource provider to set a preference on how they want their usage handled (carry over under some condition vs phase out).

**Comment by jsdelfino** [comment]
Targeting 0.0.3 for this as I agree with @hsiliev that 'somehow this feels against the idea of time-based usage'. I believe that @piotrprzybylski is OK with the current workaround to submit a monthly usage 'ping' for now, but we need to fix this in a near future after the 0.0.2 final release.

**Comment by hsiliev** [comment]
What about adding the carry-over preference into the plan? In this way the usage will remain unchanged since it represents the raw data. Some plans have time-based formulas that require the data to be carried over to the next month. It is natural to add the preference in the plan, close to the formulas. This will make things easier for plan developers and resource providers. 

Do you think this will work?

Another thing to consider is what kind of field do we want to use - simple true/false trigger or more complex value like the time windows? Do we have other use-cases besides the cf-bridge time-based formulas?

**Comment by hsiliev** [comment]
We can carry over the time-based usage by copying DB docs or resubmitting the usage to imitate the ping the resource providers have to do now. Both approaches however cause a lot of load in terms of cpu/memory & network, especially when we have a lot of entries to deal with.

I wondered if we can do this in the end of the pipeline (and not in the begining) and how complex this would be.

**Comment by hsiliev** [comment]
Filed a PR https://github.com/cloudfoundry-incubator/cf-abacus/pull/320. It adds a metric type as required field.

This will force all metering plans to add such type. To avoid this we might make this optional.

**Comment by jsdelfino** [comment]
See discussion of this issue on cf-dev:
https://lists.cloudfoundry.org/archives/list/cf-dev@lists.cloudfoundry.org/thread/6PITAHWPUVCCH4DKVYLUO6YVNA7YGZF4/

**Comment by jsdelfino** [comment]
Following up: I did another review of the design yesterday and I'm wondering about how the design we discussed (writing the usage to a single DB so we don't have to reset it every month) is going to allow usage aggregation, since the usage aggregator service and any consumers downstream the Abacus pipeline won't get a notification that usage should be considered for these time-based resources...

**Comment by jsdelfino** [comment]
So I'm thinking that we might be better off generating one event for each running app, container or whatever resource uses a time-based consumption model every month, to 'reset' its usage for the month and have it included in all the downstream databases, reports etc.

We wouldn't require resource providers, or even the CF bridge for CF runtimes, to generate these events as that'd be too much of a burden like we were discussing at the top of this thread. We'd have a batch process part of Abacus do that instead.

Looking at the traffic that would generate for runtime/apps, that'd be only one event per month and per running app. Assuming that a running app uses 5 services that generate one usage doc per 15 min, we're comparing one event per month to reset its runtime usage with about 14600 events for the services that app uses... so IMO that'll be negligible. 

Recently we've also discussed cases where we'll need to 'reset' usage counters for resource instances impacted by a metering plan, rating plan or pricing plan change. Similarly, we could generate one event for the resources impacted by a plan change, to get both usage before the change and after the change in different buckets in all the usage databases and aggregated usage reports.

So in summary I'm starting to see a common pattern here. To cover different billing periods, to reset ongoing usage at the beginning of a month, or to capture changes in the configuration governing the usage metering and rating, we just generate one event to capture the new state of a resource instance.  

Thoughts?

**Comment by hsiliev** [comment]
@jsdelfino We discussed storing the usage in both the historical and time-based DBs. Wouldn't this solve the problem with the notification?

Nevertheless if we need the reset/resubmit logic for other use-cases it seems logical to re-use the code for solving this issue as well.

**Comment by jsdelfino** [comment]
@hsiliev, @KRuelY, @piotrprzybylski, @rajkiranrbala ... Storing the usage in both DBs still wouldn't help much. For example if you had stored usage for an app in both the regular monthly-partitioned DB for May and the single 'state' DB, you would still not get that usage in the aggregated usage for June (because it would never have gone through the Abacus pipeline in June).

So the simplest way to get that usage in all the DBs for June (and not be able to notify any consumers at the end of the Abacus pipeline that this usage should be included again for June) is to inject it through the collector again, at some point in June...

The idea is then to have some kind of batch process provided by Abacus do that automatic resubmission, to not place that burden on the service providers or the CF runtime usage bridges.

Thoughts?

**Comment by jsdelfino** [comment]
I just made a small change to the keys generated for by the usage collector service to make the automatic resubmission I've described above easier to implement. See 6402cc4 for the change.

**Comment by hsiliev** [comment]
@jsdelfino: I see. So we cover the case, where consumers that expect the usage to be pushed to them, instead of pulling every now and then.

My plan would be to:
* **(metering schema, provisioning plugin)** Add a metric type, so that a  plan can have both discrete and time-based metrics. Much like https://github.com/cloudfoundry-incubator/cf-abacus/pull/320/commits
* **(account plugin, metering config)** extend account plugin to return (and cache) plan type together with plan id so the previous function can determine the type of the metrics/measures and decide which DB should be used
* **(new module)** Add a standalone app that on the 1st day of every month resubmits docs that have a time-based metric. Since we might miss late usage, retry after the slack window.

I'm absolutely sure I'm missing lots of things since the points above grade from concrete implementation to just a design idea :)

**Comment by hsiliev** [comment]
We implemented the renewer to solve this issue, so I'll close it.

There are some ideas on how to generalize the solution for all time-based events in #318
